🎉 [New Feature]: Self-elevation support

Dragon1573 opened the pull request:
## Description

- Resolve #2

## Additional information

> [!WARNING]
>
> Most modifications are manually copied from lehni/vscode-fix-checksums#1. Ideas, which originally in [spikespaz-contrib@master](https://github.com/spikespaz-contrib/vscode-fix-checksums/tree/master), **DOES NOT** have an open source license.
>
> I can't sure if this kind of "manually combining" is accepted. If this violates some rules, please tell me and reject this pull request. ❤️

Dragon1573 committed:
pnpm: Introduce 2 dependencies

- `sudo-prompt` for asking elevation
- `tmp` for creating temporary files

Signed-off-by: Dragon1573 <49941141+Dragon1573@users.noreply.github.com>

Dragon1573 committed:
New Feature: Self elevation

- Extension now can elevate to administrator privilege by itself.
- No need to launch VSCode as administrator manually.
- Automatically reload window for taking effects.

Signed-off-by: Dragon1573 <49941141+Dragon1573@users.noreply.github.com>

RimuruChan commented on extension.js:
```diff
@@ -39,31 +41,38 @@ function apply() {
     message = messages.error
   }
-  vscode.window.showInformationMessage(message)
+  if (reload) reloadWindow(message)
```
Reloading window doesn't take effects. You have to restart the process.

Dragon1573 replied:
It seems _VSCode_ does not provide an API or built-in command for restarting itself. Simply shows a message for asking manual restart.

RimuruChan commented on extension.js:
```diff
@@ -82,6 +91,70 @@ function cleanupOrigFiles() {
+      else fs.writeFile(tmpPath, writeString, encoding, err => {
+        if (err) reject(err)
+        else sudo.exec(
+          (process.platform === "win32" ? "copy /y " : "cp -f ") + `"${tmpPath}" "${filePath}"`,
```
Perhaps we can use a more general solution.
https://github.com/microsoft/vscode/blob/main/src/vs/platform/native/electron-main/nativeHostMainService.ts#L576

Dragon1573 replied:
It's not exposed to extension developers by `@types/vscode`. I think we might not able to invoke it in our code. 😞

Asynchronous function [`writeElevated`](https://github.com/microsoft/vscode/blob/eb9e2d8bcfae7dee215ec584b93a392140660661/src/vs/platform/native/electron-main/nativeHostMainService.ts#L576-L617) also uses `sudo-prompt` for asking elevation, which is what we are imitating. [`saveContent`](https://github.com/shalldie/vscode-background/blob/91d3803988b6faa0e8878c17289ba5f40eea3de1/src/background/CssFile.ts#L93-L129) in `shalldie/vscode-background` also use similar implementation for "patching" custom CSS files to VSCode core zone.

Dragon1573 committed:
Apply suggesstion from code review

- Restart VSCode itself is required.
- VSCode does not provide such an API, just ask for a manual restart.

Signed-off-by: Dragon1573 <49941141+Dragon1573@users.noreply.github.com>

Dragon1573 commented:
@RimuruChan - Thanks for acceptance ❤️

Dragon1573 commented on package.json:
```diff
@@ -51,5 +51,10 @@
     "esbuild": "^0.24.0",
     "eslint": "^9.12.0"
+  },
+  "packageManager": "pnpm@9.12.1",
```
Accidentally introduce `pnpm` as a package manager. This may encounter problem when running `pnpm exec vsce package` or `npm install`.

If there're any errors, please drop this line. ❤️

Dragon1573 commented on pnpm-lock.yaml:
Accidentally introduce `pnpm` as a package manager. This may encounter problem when running `pnpm exec vsce package` or `npm install`.

If there're any errors, please remove this file and use `npm` instead. ❤️

Dragon1573 reviewed (commented):
Accidentally introduce another package manager (`pnpm` here). Please revert to `npm` if there're any errors.

zent1n0 commented:
I'd prefer `pnpm` as package manager :blush:
